text/0000-terminology-and-boundaries.md

```diff
@@ -0,0 +1,224 @@
+- The official Nebulet kernel.
+- NVIDIA graphics drivers that implement the GPGPU firmament specification.
+- Ethernet drivers that implement the Data Plane firmament specification.
+- An implementation of the Sockets firmament.
+- TensorFlow for clusters ported to a Nebulet application that uses the GPGPU
+  and Socket firmaments.
+- Jupyter and the entire SciPy stack (lol) ported to Nebulet, using the Sockets
+  firmament and many ad-hoc application-layer IPC interfaces.
+
+Note that in all examples, we can re-use the Nebulet kernel and firmament
+interfaces.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The nucleus must provide a thin set of abstractions over the hardware, in
+particular the CPU. These abstractions must include:
+- Access and control over CPU power states, if applicable.
+- Access and control over CPU interrupts, if applicable.
+- Access and control of the MMU, if applicable.
+- Access and control over CPU registers if necessary for common drivers on
+  the target hardware.
+
+The nucleus can also implement part of a verified boot chain if such a feature
+is desirable. In this case, the nucleus will be verified by the bootloader and
+it will verify the Nebulet kernel and other startup data, such as initial
+ramdisk or commandline.
+
+The Nebulet kernel will use these functions to implement the core functionality
+of threading, scheduling, process isolation, and inter-process communication.
+
+A firmament specification will describe the IPC interfaces and datatypes that
+an application using that piece of hardware will interact with. The Data Plane
+firmament specification used in the above examples might be defined as follows
+(using Rust-like syntax):
+
+```rust
+// The supporting structs and enums are assumed to have
+// functions that can be called easily from the receiver side.
+// We'll need to flesh out the IPC system for this to be in any
+// way realistic.
+struct MacAddr([u8; 6]);
+enum Ethertype {
+    Arp,
+    Ipv4,
+    Ipv6,
+}
+
+struct EthernetPacket {
+    data: Bytes,
+}
+
+impl EthernetPacket {
+    pub fn dst(&self) -> &MacAddr;
+    pub fn src(&self) -> &MacAddr;
+    pub fn ethertype(&self) -> &Ethertype;
+    pub fn next_layer(&self) -> &L3Packet;
+}
+
+interface EthernetNic {
+    fn send(&mut self, pkt: EthernetPacket) -> Result<()>;
+    fn send_inplace(&mut self, func: impl Fn(&mut EthernetPacket) -> Result<()>) -> Result<()>;
+    fn recv(&mut self) -> Future<Result<EthernetPacket>>;
+}
+```
+
+An interface is implemented by the driver side and used by the application.
+Firmament interfaces could be provided to the application on initialization,
+like in [CloudABI](https://cloudabi.org), or they could be dynamically queried,
+depending on the security model, which is yet to be specified.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Implementing all drivers as WebAssembly applications may not be able to provide
```

Comment:
This is an important thing to think about. I'm leaning towards being able to dynamically load (or at startup only, depending on the security model) clif ir for drivers without sandboxing overhead, but this remains to be seen.

Reply:
As someone with both a security and networking background, I would argue in favor of statically-compiled drivers, with the nucleus implementing the firmament. That said, I wonder if we can somehow have it both ways.

Reply:
Okay, I'm okay with that. I suppose compiling the nucleus for different usecase is fine.